Updated Soulsby Fix as per Johan's review

sgpearson17 opened the pull request.

sgpearson17 committed:
Update section title in output documentation

Changed the section heading from 'Simulation Output' to 'Output' in docs/user/output.md for consistency and brevity.

sgpearson17 committed:
Extend Soulsby method parameters in population schema

Added new parameters to the Soulsby method configuration in population.schema.json, including flow_field_name, background_grain_size, soulsby_b_e, soulsby_theta_s, soulsby_gamma_e, soulsby_mu_d, and soulsby_freedom_factor. Updated descriptions for legacy parameters f and r to clarify their status. Will probably get rid of these later

sgpearson17 committed:
Add transport probability method to particle exposure logic

Updated ParticlePopulation to determine particle exposure based on the selected transport_probability method. The update_position method now accepts a transport_probability_method argument, and Simulation passes the configured method when updating particle positions.

sgpearson17 committed:
Add debug script for SedTRAILS simulation

Introduces debug_sedtrails.py, a script to facilitate debugging of SedTRAILS simulations in VS Code, because Stuart was having a hard time doing it from the command line. The script allows step-through debugging, breakpoint setting, and targeted testing of configuration loading and simulation object creation.

sgpearson17 committed:
Clarify distance metric as per output timestep

Updated plot title and comments to specify that the average distance is calculated per output timestep, improving clarity in the simulation dashboard.

sgpearson17 committed:
Improve plot saving logic and clarify variable names

Refactored variable names in read_netcdf for clarity and improved plot_trajectories to better determine the output directory when saving plots. The CLI now only displays a message when the plot is shown, not when saved.

sgpearson17 committed:
Remove legacy Soulsby parameters and add tracer_grain_size

Removed the legacy 'f' and 'r' parameters from the Soulsby method configuration and introduced a new 'tracer_grain_size' property to specify the grain size of tracer sediment.

sgpearson17 committed:
Improve output directory resolution in Simulation

The _get_output_dir method now resolves default or relative output directories relative to the config file's location, ensuring outputs are placed consistently regardless of the working directory. This change improves usability when running simulations from different paths.

sgpearson17 committed:
Use tracer grain size for Soulsby method and add checks

Updated PhysicsConverter and Soulsby physics plugin to use 'tracer_grain_size' instead of 'grain_diameter' when the Soulsby method is selected. Added validation for missing 'tracer_grain_size' in configuration and included comments to vectorize nested loops for performance improvements.

sgpearson17 committed:
Stuart's debugger script

local testing and debugging

sgpearson17 committed:
Update Soulsby example config with new sand populations

Renamed and updated particle populations to 'FineSand' and 'MediumSand' with revised grain sizes, tracer parameters, and seeding strategies. Adjusted transport probability to 'no_probability' and modified bounding boxes and number of locations for random seeding. Commented out the output directory path.

sgpearson17 committed:
reverted unnecessary change to cli main

reverted unnecessary changes as per Johan's PR review, and confirmed that the code still works for both saving and not saving. this was a copilot goof, sorry!

sgpearson17 committed:
Remove commented-out is_exposed calculation

As per Johan's PR review, deleted unused code related to the calculation of the 'is_exposed' particle property to clean up the ParticlePopulation class.

sgpearson17 committed:
Add docstring for transport_probability_method param

Updated as per Johan PR review. Updated the ParticlePopulation class docstring to include the transport_probability_method parameter, clarifying its purpose in the method signature.

sgpearson17 committed:
Refactor output_dir handling to use Path objects

Changed as per Johan PR review. Updated DataManager and Simulation classes to accept and use pathlib.Path objects for output_dir instead of plain strings. This improves path handling consistency and type safety throughout the codebase.

sgpearson17 commented:
@jreyns I think I screwed something up because I thought I was fixing the other PR but instead it made a new one. If you want to discuss just let me know...

jreyns commented on src/sedtrails/particle_tracer/particle_seeder.py:
```diff
@@ -477,7 +482,7 @@ def update_status(self) -> None:
         )
 
-    def update_position(self, flow_field: Dict, current_timestep: float) -> None:
+    def update_position(self, flow_field: Dict, current_timestep: float, transport_probability_method: str = None) -> None:
```
Not used, remove

sgpearson17 replied:
removed

jreyns commented on src/sedtrails/particle_tracer/particle_seeder.py:
```diff
@@ -487,6 +492,8 @@ def update_position(self, flow_field: Dict, current_timestep: float) -> None:
             A dictionary containing the flow field information.
         current_timestep : float
             The current time step in the simulation in seconds.
+        transport_probability_method: str
```
not used, remove

sgpearson17 replied:
removed

jreyns commented on src/sedtrails/pathway_visualizer/trajectories.py:
```diff
@@ -8,12 +8,12 @@
 
 
-def read_netcdf(file_path):
+def read_netcdf(results_file_path):
```
add typing to function argument

sgpearson17 replied:
what do you mean by this?

sgpearson17 replied:
like ```def read_netcdf(results_file_path: Union[str, Path]) -> xr.Dataset:```? this is what copilot suggests

sgpearson17 replied:
ah but it should just be a path and not a string, right?

sgpearson17 replied:
```def read_netcdf(results_file_path: Path) -> xr.Dataset:``` ?

jreyns commented on src/sedtrails/simulation_orchestrator/simulation_manager.py:
```diff
@@ -75,7 +76,7 @@ def __init__(self, config_file: str, enable_dashboard: Optional[bool] = None):
         self.writer = self.data_manager.writer
 
-        setup_logging(output_dir=self.writer.output_dir)  # Initialize logging in the results directory
+        setup_logging(output_dir=str(self.writer.output_dir))  # Initialize logging in the results directory
```
is this returning something?

sgpearson17 replied:
I don't think so, as far as I can tell it just chooses where to put the log file.

jreyns commented on src/sedtrails/simulation_orchestrator/simulation_manager.py:
```diff
@@ -505,7 +533,7 @@ def run(self):
         # from sedtrails.data_manager.simulation_netcdf_writer import SimulationNetCDFWriter
 
-        # writer = SimulationNetCDFWriter(self._get_output_dir())
+        # writer = SimulationNetCDFWriter(str(self._get_output_dir()))
```
Path?

sgpearson17 replied:
this part was all commented out so I just removed it

jreyns commented on src/sedtrails/transport_converter/physics_converter.py:
```diff
@@ -109,13 +109,23 @@ def grain_properties(self):
-            self.config.water_density,
-            self.config.kinematic_viscosity,
-        )
+        if self.config.tracer_method == 'soulsby':
```
I would make this a match case construct, so it allows easy implementation of future transport methods

sgpearson17 replied:
what do you mean? like to match if it is soulsby/Soulsby/SOULSBY?

jreyns replied:
there will be other tracer methods in the future, so it is nice to have a match construct here already.

jreyns commented on src/sedtrails/transport_converter/plugins/physics/soulsby.py:
```diff
@@ -45,7 +45,9 @@ def add_physics(self, sedtrails_data: SedtrailsData, grain_properties: dict[str,
-        grain_size = self.config.grain_diameter
+        grain_size = self.config.tracer_grain_size
+        if grain_size is None:
+            raise ValueError("Missing required 'tracer_grain_size' in configuration.")
```
shouldn't our validator catch that?

sgpearson17 replied:
yes true. I removed the field from the config file and our validator caught it. I removed this check and the other similar ones from soulsby.py

jreyns reviewed (changes requested):
Some more remarks, sorry

sgpearson17 committed:
corrected Bertin citation

to 2008 paper

sgpearson17 committed:
corrected soulsby references

previous refs were incorrect

sgpearson17 committed:
Remove redundant config validation in Soulsby plugin

Eliminated explicit checks for required grain property values in the Soulsby physics plugin  because this is already handled in validator. This streamlines property extraction and assumes upstream validation or handling of missing values.

sgpearson17 committed:
Refactor exposure logic and update method signature

Added comments clarifying exposure criteria for van Westen and Soulsby methods. Removed unused 'transport_probability_method' parameter from 'update_position' method signature for clarity.

sgpearson17 committed:
Update type hints for read_netcdf function

Changed the parameter type of results_file_path in read_netcdf from str or Path to Path, and added a return type hint for xr.Dataset to improve type safety and code clarity.

sgpearson17 committed:
Remove unused transport_probability argument

Eliminates the 'transport_probability' argument from the population.update_position call, as it is no longer required. Also removes commented-out code related to SimulationNetCDFWriter for clarity.

sgpearson17 committed:
modified debug_sedtrails.py

changed filepaths on Stuart's machine for systematic testing

sgpearson17 committed:
Merge branch 'sedtrails:dev' into soulsby-fix

sgpearson17 committed:
fixed ruff errors

Eliminated an unused import from trajectories.py and commented out an unused variable assignment in soulsby.py to clean up the code.

sgpearson17 committed:
Merge branch 'soulsby-fix' of https://github.com/sgpearson17/sedtrails-sgp into soulsby-fix

sgpearson17 committed:
Add validation for required grain properties in Soulsby plugin

Raises ValueError if critical_shields, dimensionless_grain_size, or settling_velocity are missing from grain_properties, ensuring necessary inputs for Soulsby physics calculations.

sgpearson17 committed:
Add comprehensive unit tests for physics_lib.py

Introduces a new test suite covering all major functions in sedtrails.transport_converter.physics_lib. Tests include grain property calculations, flow dynamics, transport mechanisms, layer thickness computations, vector operations, edge cases, and integration workflows, ensuring correctness and robustness across a wide range of sediment transport scenarios.

sgpearson17 committed:
Add documentation for unit testing physics_lib.py

Introduces a guide for adding unit tests to physics_lib.py, including test structure, categories, examples, and best practices. This helps developers write effective tests for new features and maintain code reliability.

sgpearson17 committed:
Add unit test documentation to developer guide

Introduces a new 'Adding Unit Tests' section to the documentation and updates contribution guidelines to reference it. Also corrects numbering in contribution steps and includes the new section in the main documentation index.

sgpearson17 commented:
@jreyns I added some unit tests for physics_lib.py. There is still a lot more that could be done, and I think the soulsby.py code should be consolidated further by only using common elements of physics_lib.py where applicable, but I think it would be good also for @the-vdag 's sake to have this improved version of Soulsby in the dev branch, so if you could review my changes soon that would be great. Thanks!

sgpearson17 committed:
Merge branch 'dev' into soulsby-fix

jreyns commented on src/sedtrails/data_manager/manager.py:
```diff
@@ -53,10 +55,17 @@ def __init__(self, output_dir: str, max_bytes=512 * 1024 * 1024):
         ----------
         
-        output_dir : str
+        output_dir : Path
```
str or Path

jreyns reviewed (approved):
will approve for now, so we can get going

sgpearson17 commented:
Gracias senor!
